src/Shared/FileMatcher.cs

```diff
@@ -1782,9 +1782,27 @@ internal string[] GetFiles
 
         private static string ComputeFileEnumerationCacheKey(string projectDirectoryUnescaped, string filespecUnescaped, IEnumerable<string> excludes)
         {
+            Debug.Assert(projectDirectoryUnescaped != null);
+            Debug.Assert(filespecUnescaped != null);
+
             var sb = new StringBuilder();
 
-            sb.Append(projectDirectoryUnescaped);
+            if (filespecUnescaped.Contains(".."))
+            {
+                filespecUnescaped = FileUtilities.GetFullPathNoThrow(filespecUnescaped);
+            }
+
+            var filespecIsAnAbsoluteGlobPointingOutsideOfProjectCone =
+                Path.IsPathRooted(filespecUnescaped) &&
+                !filespecUnescaped.StartsWith(projectDirectoryUnescaped, StringComparison.OrdinalIgnoreCase);
+
+            // If we include the project directory when the filespec does not depend on it we'll get cache misses
+            // when the project directory independent glob repeats for each project.
```

Comment:
I get the gist of this comment but I found it a bit confusing. I think it'd be clearer if you said `when the project-directory-independent glob is used in multiple projects`?